// nylas/lib/config.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var webhook_1 = require("./models/webhook");
exports.apiServer = null;
function setApiServer(newApiServer) {
    exports.apiServer = newApiServer;
}
exports.setApiServer = setApiServer;
exports.clientSecret = '';
function setClientSecret(newClientSecret) {
    exports.clientSecret = newClientSecret;
}
exports.setClientSecret = setClientSecret;
exports.timeout = 0;
function setTimeout(newTimeout) {
    exports.timeout = newTimeout;
}
exports.setTimeout = setTimeout;
var ResponseType;
(function (ResponseType) {
    ResponseType["CODE"] = "code";
    ResponseType["TOKEN"] = "token";
})(ResponseType = exports.ResponseType || (exports.ResponseType = {}));
var Region;
(function (Region) {
    Region["Us"] = "us";
    Region["Ireland"] = "ireland";
})(Region = exports.Region || (exports.Region = {}));
exports.DEFAULT_REGION = Region.Us;
exports.regionConfig = (_a = {},
    _a[Region.Us] = {
        nylasAPIUrl: 'https://api.nylas.com',
        dashboardApiUrl: 'https://dashboard-api.nylas.com',
        callbackDomain: 'cb.nylas.com',
        websocketDomain: 'tunnel.nylas.com',
        telemetryApiUrl: 'https://cli.nylas.com',
    },
    _a[Region.Ireland] = {
        nylasAPIUrl: 'https://ireland.api.nylas.com',
        dashboardApiUrl: 'https://ireland.dashboard.nylas.com',
        callbackDomain: 'cb.nylas.com',
        websocketDomain: 'tunnel.nylas.com',
        telemetryApiUrl: 'https://cli.nylas.com',
    },
    _a);
exports.DEFAULT_WEBHOOK_TRIGGERS = Object.values(webhook_1.WebhookTriggers);
